### PR TITLE
feat(aws-ecs): support runtime platform property for create fargate windows runtime.

### LINKED ISSUE #17242: (aws-ecs): support RuntimePlatform Property at TaskDefinition class for create windows OperatingSystem
### Description

support `RuntimePlatform` Property at `TaskDefinition` class for create windows OperatingSystem.

### Use Case

Windows support for AWS Fargate on Amazon ECS on 28 OCT 2021 .
https://aws.amazon.com/tw/blogs/containers/running-windows-containers-with-amazon-ecs-on-aws-fargate/

### Proposed Solution

support RuntimePlatform Property at TaskDefinition class.

### Other information

_No response_

### Acknowledge

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by neilkuan**
feat(aws-ecs): support runtime platform property for create fargate windows and Graviton2 runtime.

close #17242
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by neilkuan** [comment]
Hi @upparekh ,If you have time, please take a look at this PR, thanks. ☺️

**Comment by pahud** [comment]
Hi @kimisme9386 @clarencetw @dwchiang 

Are you interested to take an initial review? Please leave a comment with "LGTM" if you think it's good with you.

Thanks.

**Comment by pahud** [comment]
@neilkuan Any chance we can add this new feature into this PR? 

Announcing AWS Graviton2 Support for AWS Fargate – Get up to 40% Better Price-Performance for Your Serverless Containers
https://aws.amazon.com/tw/blogs/aws/announcing-aws-graviton2-support-for-aws-fargate-get-up-to-40-better-price-performance-for-your-serverless-containers/

**Comment by neilkuan** [comment]
> @neilkuan Any chance we can add this new feature into this PR?
> 
> Announcing AWS Graviton2 Support for AWS Fargate – Get up to 40% Better Price-Performance for Your Serverless Containers https://aws.amazon.com/tw/blogs/aws/announcing-aws-graviton2-support-for-aws-fargate-get-up-to-40-better-price-performance-for-your-serverless-containers/

yes, I will, wip.

**Review by kimisme9386 — COMMENTED** [review]
LGTM

**Review by clarencetw — COMMENTED** [review]
LGTM

**Comment by dwchiang** [comment]
LGTM. It's great to have Graviton2 for AWS Fargate.

**Review by dwchiang — COMMENTED** [review]
LGTM. It's great to have Graviton2 for AWS Fargate.

File: packages/@aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts:149]: This makes an assumption that all non-Linux operatingSystemFamily values will all be windows. Is that the case? Or would it be possible that some non-Linux and non-Windows operatingSystemFamily values will be added to this? 

I think we probably need to check that the operatingSystemFamily is a WINDOWS_XXX one, rather than checking that it is *not* LINUX.

File: packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:48]: Can you add the new property to this test? the test is called 'with all properties set'

File: packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:315]: For readability:
```suggestion
    test('creating a Fargate TaskDefinition with WINDOWS_SERVER_X operatingSystemFamily and incorrect cpu throws an error', () => {
```

File: packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:260]: Can you separate the Graviton part to it's own test?

**Review by madeline-k — CHANGES_REQUESTED** [review]
Thanks for opening this PR, @neilkuan! I have a few questions/comments.

File: packages/@aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts
Comment (neilkuan) [reply packages/@aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts:150]: Yes, you are right, @pahud  wdyt?

> It's strange to me that this validation happens only in `FargateTaskDefinition` and not `TaskDefinition`. I see the same thing is happening with the ephemeral storage validation. But it seems like a bad customer experience to get this synth-time feedback only if you are using the `FargateTaskDefinition` class and not with `TaskDefinition`.
> 
> Am I understanding correctly that this is the case? It seems like we should move this up to the `TaskDefinition` constructor.

File: packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts
Comment (neilkuan) [reply packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:260]: okay, no problem. 👍

File: packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts
Comment (neilkuan) [reply packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts:217]: It is also not stated clearly in the cloudformation document. 
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-taskdefinition.html#cfn-ecs-taskdefinition-runtimeplatform

File: packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts
Comment (pahud) [inline packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts:217]:
```suggestion
```

I think it's unclear in the cloudformation document and could be confusing as there is actually no `runtimePlatform` property for the service resource. Let's just remove it. @madeline-k  wdyt?

**Review by pahud — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts
Comment (pahud) [reply packages/@aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts]: Yap I agree we validate this in the `TaskDefinition` base class.

File: packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts
Comment (pahud) [inline packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts:437]:
```suggestion
    // validate the cpu and memory size for the Windows operation system family.
    if (props.runtimePlatform?.operatingSystemFamily?._operatingSystemFamily.includes('WINDOWS')) {
```

**Review by pahud — CHANGES_REQUESTED** [review]

**Comment by neilkuan** [comment]
Hi @madeline-k , I already made some improvement. please review at the free time, thanks.

**Comment by neilkuan** [comment]
Hi @madeline-k , I already made some improvement. please review at the free time, thanks.

File: packages/@aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts:67]: Let's just use the exact same wording as in task-definition.ts:
```suggestion
   * A runtimePlatform is supported only for tasks using the Fargate launch type.
```

File: packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts:731]: These errors should use the property names that the customer supplied. So vCPU -> cpu and Memory -> memoryMiB. Since the user has to supply the values in MiB, I think it would be easier for the message to us MiB instead of GB. And, change the wording to be more user friendly:

```suggestion
        throw new Error(`If provided cpu is ${cpu}, then memoryMiB must have a min of 1024 and a max of 8192, in 1024 increments. Provided memoryMiB was ${Number(memory)}.`);
```

File: packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts:735]:
```suggestion
        throw new Error(`If provided cpu is ${cpu}, then memoryMiB must have a min of 4096 and max of 16384, in 1024 increments. Provided memoryMiB ${Number(memory)}.`);
```

File: packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts:739]:
```suggestion
        throw new Error(`If provided cpu is ${cpu}, then memoryMiB must have a min of 8192 and a max of 30720, in 1024 increments. Provided memoryMiB was ${Number(memory)}.`);
```

File: packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/base/task-definition.ts:438]: Can you add a comment similar to this? 
```suggestion
      // We know that props.cpu and props.memoryMiB are defined because an error would have been thrown previously if they were not. But, typescript is not able to figure this out, so using the ! operator here to let the type-checker know they are defined. 
      this.checkFargateWindowsBasedTasksSize(props.cpu!, props.memoryMiB!, props.runtimePlatform!);
```

File: packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:8]: Can you add the new properties to the 'with all properties set' test?

File: packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:282]: This test has the exact same name as the previous one. Should this one be called 'runtime testing for linux container'?

File: packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:345]: Can you include one more test here that checks that the right error is thrown if `cpu` and `memoryLimitMiB` are both undefined, but `runtimePlatform` is? I believe the error should be: `Fargate-compatible tasks require both CPU (${props.cpu}) and memory (${props.memoryMiB}) specifications`

**Review by madeline-k — CHANGES_REQUESTED** [review]
Thanks for making updates, @neilkuan! This is looking great, I just have a few comments about wording and unit tests.

File: packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts
Comment (neilkuan) [reply packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:345]: Hi @madeline-k , I found this source code.
https://github.com/aws/aws-cdk/blob/cac11bba2ea0714dec8e23b069496d1b9d940685/packages/%40aws-cdk/aws-ecs/lib/fargate/fargate-task-definition.ts#L124-L131

if `props.cpu` or `props.memoryMiB` is undefined, I think it will been defined to default value right?!

wdyt @pahud

Comment (madeline-k) [reply packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:362]: You're right, if someone creates a new FargateTaskDefinition, there will be those default values set. But if they create a new TaskDefinition, and specify a runtimePlatform, but no memory and cpu, that should cause an error. You can add a test to `task-definition.test.ts`. But I can see the argument that that is out of scope for this PR, so up to you. If you decide to opt to not include that, let me know and I can approve. Thanks for making these latest updates!

Comment (neilkuan) [reply packages/@aws-cdk/aws-ecs/test/fargate/fargate-task-definition.test.ts:362]: Hi @madeline-k,thanks for review. 
I will create another PR for that testing. 

Let's let this PR approved.😆

**Comment by neilkuan** [comment]
Hi @madeline-k ready to merge. 😁

**Review by madeline-k — DISMISSED** [review]
😄

**Review by madeline-k — APPROVED** [review]
Approving, my review was dismissed after merging from master.

**Pull request merged**
